[5.5.x] Update disk check

bernardjkim opened the pull request:
### Description
This PR updates the disk check to report two separate warning and critical probes at different thresholds.

### Linked tickets and PRs
* Ports https://github.com/gravitational/satellite/pull/224

bernardjkim committed:
Add a warning watermark to the disk check

- Set default critical watermark to 90%

bernardjkim committed:
Bump protobuf to 1.3.0 to meet gravity constraint

bernardjkim committed:
Warning and critical thresholds configured separately

a-palchikov commented on monitoring/storage.go:
```diff
@@ -34,14 +36,43 @@ type StorageConfig struct {
+	}
+
+	if c.WatermarkCritical > 100 {
+		errors = append(errors, trace.BadParameter("critical watermark must be between 0 and 100"))
```
Do you want to make sure Critical > Warning?

bernardjkim replied:
The ui might be a little weird if we ensure that critical > warning. Example: if we set critical to 50 and leave warning unset, warning will default to 80. 

I guess what we can do is first set defaults. Then, if warning > critical just set warning == critical?

a-palchikov commented on monitoring/storage.go:
```diff
@@ -50,17 +81,29 @@ type HighWatermarkCheckerData struct {
-func (d HighWatermarkCheckerData) FailureMessage() string {
+// WarningMessage returns warning watermark check message
+func (d HighWatermarkCheckerData) WarningMessage() string {
+	return fmt.Sprintf("disk utilization on %s exceeds %v percent (%s is available out of %s), see https://gravitational.com/telekube/docs/cluster/#garbage-collection",
```
I would probably mention that the cluster will degrade once it reaches %v (critical watermark).

r0mant replied:
I think this link needs updating as there's no more telekube: https://gravitational.com/telekube/docs/cluster/#garbage-collection.

The current link is https://gravitational.com/gravity/docs/cluster/#garbage-collection. Same below.

a-palchikov commented on monitoring/storage_linux.go:
```diff
@@ -143,8 +147,13 @@ func (c *storageChecker) checkFsType(ctx context.Context, reporter health.Report
 
+// checkHighWatermark checks the disk usage. A failed warning or critical probe
+// will be reported if the usage percentage is above the set thresholds.
+// If the WatermarkWarning percentage is higher than or equal to the
```
Does it make sense to be able to set warning values exceeding the critical value?

a-palchikov commented on monitoring/storage_linux.go:
```diff
@@ -143,8 +147,13 @@ func (c *storageChecker) checkFsType(ctx context.Context, reporter health.Report
+// If the WatermarkWarning percentage is higher than or equal to the
+// WatermarkCritical percentage, then the check will only ever report critical
+// probes.
 func (c *storageChecker) checkHighWatermark(ctx context.Context, reporter health.Reporter) error {
```
I'd call this `checkDiskUsage` or something similar to emphasize the purpose.

a-palchikov commented on monitoring/storage_linux_test.go:
```diff
@@ -103,18 +103,20 @@ func (_ *StorageSuite) TestStorage(c *C) {
-			HighWatermark: 40,
+			Path:              path.Join("/tmp", fmt.Sprintf("%d", time.Now().Unix())),
+			WillBeCreated:     true,
+			WatermarkCritical: 40,
```
Not sure these values are useful. What about exercising happy path with warning < critical and disk usage reaching both warning, then criticial. Also warning >= critical which should yield an error, imo.

r0mant commented on monitoring/storage.go:
```diff
@@ -34,14 +36,43 @@ type StorageConfig struct {
+	WatermarkCritical uint
+}
+
+func (c *StorageConfig) CheckAndSetDefaults() error {
```
Nit: Missing godoc.

r0mant reviewed (commented):
lgtm with Dima's comments

bernardjkim committed:
Check warning watermark <= critical watermark

bernardjkim committed:
Update messages

- Warning message indicates when usage will exceed critical watermark
- Update garbage collection link

bernardjkim committed:
Update godocs

- checkHighWatermark -> checkDiskUsage

bernardjkim committed:
Warning/Critical -> Low/High

bernardjkim committed:
Update test cases

a-palchikov commented on monitoring/storage_linux.go:
```diff
@@ -147,15 +147,16 @@ func (c *storageChecker) checkFsType(ctx context.Context, reporter health.Report
+	if c.HighWatermark == 0 {
 		return nil
 	}
+	if c.LowWatermark > c.HighWatermark {
```
I think this is redundant as already being validated in `StorageConfig.CheckAndSetDefaults`?

a-palchikov commented on monitoring/storage.go:
```diff
@@ -36,43 +36,50 @@ type StorageConfig struct {
+	}
+
+	if c.LowWatermark > c.HighWatermark {
+		errors = append(errors, trace.BadParameter("low watermark (%v) cannot be higher than high watermark (%v)", c.LowWatermark, c.HighWatermark))
```
This is going to be an issue unless you always set both values in a single go. Would it make sense to have defaults be relative to each other like default low watermark is always 10% smaller than the high watermark?

bernardjkim replied:
https://github.com/gravitational/satellite/pull/224#discussion_r436213986
I think we wanted to be able to configure the two values separately.

Ok, so instead of returning an error if LowWatermark > HighWatermark, we'll just set LowWatermark = HighWatermark. This way LowWatermark <= HighWatermark.

bernardjkim committed:
LowWatermark is always <= HighWatermark

r0mant commented on monitoring/storage.go:
```diff
@@ -50,9 +88,15 @@ type HighWatermarkCheckerData struct {
-	return fmt.Sprintf("disk utilization on %s exceeds %v percent (%s is available out of %s), see https://gravitational.com/telekube/docs/cluster/#garbage-collection",
+// WarningMessage returns warning watermark check message
+func (d HighWatermarkCheckerData) WarningMessage() string {
+	return fmt.Sprintf("disk utilization on %s exceeds %v percent, cluster will degrade if usage exceeds %v percent (%s is available out of %s), see https://gravitational.com/gravity/docs/cluster/#garbage-collection",
```
Nit: I'd move `(%s is available out of %s)` after the first part of the sentence, otherwise it reads a bit confusing:

```
disk utilization on %s exceeds %v percent (%s is available out of %s), cluster will degrade if usage exceeds %v percent, see https://gravitational.com/gravity/docs/cluster/#garbage-collection
```

r0mant commented on Gopkg.toml:
```diff
@@ -29,7 +29,7 @@ required = [
 [[constraint]]
   name = "github.com/gogo/protobuf"
-  version = "=v1.2.1"
+  version = "=v1.3.0"
```
What's this dependency bump for? Just making sure we don't run into any incompatibility issues.

bernardjkim replied:
Just matching protobuf version in gravity https://github.com/gravitational/gravity/blob/95d4167587056e66cfecc55e391f4bff29d55481/Gopkg.toml#L74

r0mant reviewed (approved):
Much cleaner now 👍

bernardjkim committed:
Edit message

- percent -> %
